### PR TITLE
Display jurisdiction Names instead of jurisdiction ids in updatePlan 

### PR DESCRIPTION
**Pull request by p-netm**
part of #417 

- Adds new JurisdictionIdsToNames component that takes in a string array of ids and lazy loads the names either from store or from the the opensrp service.

### CODE REVIEW

File: src/components/JurisdictionIdToName/index.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+JurisdictionIdToName.defaultProps = defaultProps;
+
+export { JurisdictionIdToName };
+
+// connect component
+
+/** interface for ampState returned obj */
+interface DispatchedProps {
+  idsToNamesLookup: FlexObject<string>;
```
Comment (moshthepitt) [inline src/components/JurisdictionIdToName/index.tsx]: I think it is probably possible to not use FlexObject here - we already know what we are dealing with based on line 108, no?

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx [hunk 2]
```diff
@@ -5,6 +5,9 @@ import { connect } from 'react-redux';
 import { RouteComponentProps } from 'react-router';
 import { Col, Row } from 'reactstrap';
 import { Store } from 'redux';
+import ConnectedJurisdictionIdToName, {
+  JurisdictionIdsToNames,
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/index.tsx]: I think that this should be `JurisdictionIdToName`.  If that is the case, why is a test not failing?  What test do we need to add?

File: src/containers/forms/PlanForm/index.tsx [hunk 3]
```diff
@@ -74,6 +75,7 @@ export interface PlanFormProps {
   initialValues: PlanFormFields /** initial values for fields on the form */;
   jurisdictionLabel: string /** the label used for the jurisdiction selection */;
   redirectAfterAction: string /** the url to redirect to after form submission */;
+  JurisdictionIdToName?: typeof ConnectedJurisdictionIdToName | null;
```
Comment (moshthepitt) [inline src/containers/forms/PlanForm/index.tsx]: Can we add a docstring here to match all the other props?  For consistency :)

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx (hunk 2 above)
Comment (moshthepitt) [reply src/containers/pages/InterventionPlan/UpdatePlan/index.tsx]: Actually I think that this is an unused import.

File: src/containers/pages/InterventionPlan/UpdatePlan/tests/__snapshots__/index.test.tsx.snap [hunk 4]
```diff
@@ -82,6 +82,7 @@ exports[`components/InterventionPlan/UpdatePlan renders plan definition list cor
 
 exports[`components/InterventionPlan/UpdatePlan renders plan definition list correctly: PlanForm 1`] = `
 Object {
+  "JurisdictionIdToName": null,
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/tests/__snapshots__/index.test.tsx.snap]: Do we need another test where this is not null?

File: src/containers/pages/InterventionPlan/UpdatePlan/tests/index.test.tsx [hunk 5]
```diff
@@ -22,6 +22,7 @@ describe('components/InterventionPlan/UpdatePlan', () => {
     const mock: any = jest.fn();
     const serviceMock: any = jest.fn(async () => []);
     return {
+      JurisdictionIdToName: null,
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/tests/index.test.tsx]: Should we test when this is not null?

File: src/helpers/utils.tsx [hunk 6]
```diff
@@ -36,8 +36,8 @@ import { Plan } from '../store/ducks/plans';
 import { InitialTask } from '../store/ducks/tasks';
 import { colorMaps, ColorMapsTypes } from './structureColorMaps';
 /** Interface for an object that is allowed to have any property */
-export interface FlexObject {
-  [key: string]: any;
+export interface FlexObject<T = any> {
```
Comment (moshthepitt) [inline src/helpers/utils.tsx]: Nice

File: src/components/JurisdictionIdToName/index.tsx [hunk 7]
```diff
@@ -0,0 +1,115 @@
+/** will be responsible for lazy loading a jurisdiction id
```
Comment (moshthepitt) [inline src/components/JurisdictionIdToName/index.tsx]: This should me moved to `containers`

File: src/components/JurisdictionIdToName/index.tsx (hunk 1 above)
Comment (p-netm) [reply src/components/JurisdictionIdToName/index.tsx]: Then i think, i should remove the interface totally.
Ok, so if the types of the obj returned by `mapStateToProps` are implicitly well defined, then we dont need an interface describing the output, in this case the `DispatchedProps` interface?

**Comment by p-netm** [comment]
> Do you think that instead of editing the top level jurisdiction redux module we should create a new one inside the OpenSRP directory inside ducks? One reason for this may be ease of portability i.e. we may need to move this to OpenSRP web at some point.

I dont think i have arguments against this, so i will do this.

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx (hunk 2 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/index.tsx]: There isn't yet an integration test for JurisdictionIdToName in updatePlan. will add one.

File: src/containers/forms/PlanForm/index.tsx (hunk 3 above)
Comment (p-netm) [reply src/containers/forms/PlanForm/index.tsx]: :+1:

File: src/components/JurisdictionIdToName/index.tsx (hunk 1 above)
Comment (moshthepitt) [reply src/components/JurisdictionIdToName/index.tsx]: I had not considered not using it completely, I just noticed that we dont need to use FlexObject because we already know the shape of what we are dealing with

File: src/store/ducks/jurisdictions.ts [hunk 8]
```diff
@@ -85,10 +85,12 @@ export type ImmutableJurisdictionState = JurisdictionState &
 /** initial state */
 const initialState: ImmutableJurisdictionState = SeamlessImmutable({
   allJurisdictionIds: {} as AllJurisdictionIds,
+  geoLessJurisdictionsById: {},
```
Comment (moshthepitt) [inline src/store/ducks/jurisdictions.ts]: Do we still need this here?

Comment (p-netm) [reply src/store/ducks/jurisdictions.ts]: nope

**Comment by p-netm** [comment]
Just realized a problem in this, apparently the ids are chunked correctly but the component still makes an api request for each id,

File: src/store/ducks/opensrp/jurisdictions/index.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
+import { get, keyBy, values } from 'lodash';
```
Comment (cKellyDesign) [inline src/store/ducks/opensrp/jurisdictions/index.ts]: @p-netm what is the reasoning for not simply using the existing `src/store/ducks/jurisdictions.ts`?

File: src/containers/forms/PlanForm/index.tsx [hunk 10]
```diff
@@ -257,7 +262,16 @@ const PlanForm = (props: PlanFormProps) => {
                       <ul id="selected-jurisdiction-list">
                         {values.jurisdictions.map((jurisdiction, index) => (
```
[the remark this answers is not in the document]
Comment (p-netm) [reply src/containers/forms/PlanForm/index.tsx]: @cKellyDesign There was this piece of planform-specific code that's coupled with each of those ids,

```typescript
<fieldset>
                              <Field
                                type="hidden"
                                readOnly={true}
                                name={`jurisdictions[${index}].id`}
                                id={`jurisdictions-${index}-id`}
                                value={jurisdiction.id}
                              />
                              <Field
                                type="hidden"
                                readOnly={true}
                                name={`jurisdictions[${index}].name`}
                                id={`jurisdictions-${index}-name`}
                                value={jurisdiction.name}
                              />
</fieldset>
```

I was not sure if moving this code into JurisdictionIdToName component would have been ideal. I suppose i could move the fieldset into the `JurisdictionidToName` component. should i make this change?

File: src/store/ducks/opensrp/jurisdictions/index.ts (hunk 9 above)
Comment (p-netm) [reply src/store/ducks/opensrp/jurisdictions/index.ts]: Mosh thought this should go here, i think he mentioned something to do with portability of this ducks module to other FE apps that call to opensrp,.. i am not sure though if this was the reason, @moshthepitt, please help me answer this :point_up:

Comment (cKellyDesign) [reply src/store/ducks/opensrp/jurisdictions/index.ts]: > portability of this ducks module to other FE apps that call to opensrp

Ok yeah that makes sense 👍

File: src/containers/forms/PlanForm/index.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/containers/forms/PlanForm/index.tsx]: Oh interesting, maybe you could pass it in as the mapping function which is used in the `JurisdictionidToName` render function. Maybe something like:

```
<JurisdictionidToName
  id={values.jurisdictions.map(jur => jur.id)} 
  someMapperProp={(d, i) => (<li key={i}><fieldset>...</fieldset></li>)}
/>
```
```
// in JurisdictionidToName
const jurisdictionsArray = [...jurisdictions for this plan which have just been fetched]
const listItems = jurisdictionsArray.map(this.props.someMapperProp)
return <ul>{listItems}</ul>;
```

@moshthepitt what do you think about all this?

[the remark this answers is not in the document]
Comment (moshthepitt) [reply src/containers/forms/PlanForm/index.tsx]: @cKellyDesign I think that is a nice solution

**Comment by moshthepitt** [comment]
We probably need to use this one https://github.com/OpenSRP/opensrp-server-web/issues/139

File: src/constants.tsx [hunk 11]
```diff
@@ -367,3 +369,11 @@ export const FI_PLAN_TYPE = 'FI';
 export type FI_PLAN_TYPE = typeof FI_PLAN_TYPE;
 export const IRS_PLAN_TYPE = 'IRS';
 export type IRS_PLAN_TYPE = typeof IRS_PLAN_TYPE;
+
+//
```
Comment (cKellyDesign) [inline src/constants.tsx]: Some cleanup

File: src/constants.tsx [hunk 12]
```diff
@@ -367,3 +369,11 @@ export const FI_PLAN_TYPE = 'FI';
...
+
+/** dictates how many jurisdiction ids to pass as filter parameter to the
+ * findJurisdictionById endpoint.
+ */
+export const JURISDICTION_NAMES_CHUNK_SIZE = 17;
+export type JURISDICTION_NAMES_CHUNK_SIZE = typeof JURISDICTION_NAMES_CHUNK_SIZE;
```
Comment (cKellyDesign) [inline src/constants.tsx]: Is this something that could/should be configured differently for different implementations? If so then should go in `env.ts`, otherwise looks good here.

File: src/helpers/utils.tsx (hunk 6 above)
Comment (cKellyDesign) [reply src/helpers/utils.tsx]: Yeah nice touch

File: src/constants.tsx (hunk 12 above)
Comment (p-netm) [reply src/constants.tsx]: Yeah , its configurable

**Comment by cKellyDesign** [comment]
> It seems that our solution is failing when the number of jurisdictions is too many.

@moshthepitt it looks like this has been resolved 🎉

**Comment by moshthepitt** [comment]
@cKellyDesign @p-netm I think we still need to refactor slightly to use https://github.com/OpenSRP/opensrp-server-web/issues/139

**Comment by cKellyDesign** [comment]
@moshthepitt is OpenSRP/opensrp-server-web#139 ready for implementation?

**Comment by moshthepitt** [comment]
@cKellyDesign I think it is okay to wait for that to be ready.  I feel that it will give us a much better way to achieve the goals of this PR without the chunked repeated requests to the server, and I feel that there is not real pressure to have this issue fixed right now so we can afford to wait until the PR on the OpenSRP server side is approved and merged.  Do you agree?

**Comment by cKellyDesign** [comment]
@p-netm what's the status on this one? Do you need anything to complete?

**Comment by p-netm** [comment]
> @p-netm what's the status on this one? Do you need anything to complete?




@cKellyDesign This is done, ready for rereview and merge.

File: src/components/forms/PlanForm/index.tsx [hunk 13]
```diff
@@ -127,6 +126,14 @@ export interface PlanFormProps {
...
   jurisdictionLabel: string /** the label used for the jurisdiction selection */;
   redirectAfterAction: string /** the url to redirect to after form submission */;
+  renderLocationNames?: (
+    child?: (
+      locationName: string,
+      locationId: string,
+      index: number
+    ) => JSX.Element /** nested render content for each location name */
+  ) => JSX.Element;
+  /** a render prop that renders the plan's location names */
```
Comment (cKellyDesign) [inline src/components/forms/PlanForm/index.tsx]: @p-netm why the nested functions? Is it possible to flatten to a single func that just accepts those three params?

**Comment by p-netm** [comment]
@moshthepitt @cKellyDesign Ready for rereview

File: src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+    serve
+      .read(planId)
+      .then((response: Location[]) => {
+        store.dispatch(actionCreator(response, planId));
+      })
+      .catch((err: Error) => {
+        displayError(err);
+      });
+  }
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx]: @p-netm would it be possible to move this out of the component and export it as its own const? (for testability?)

File: src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+  useEffect(() => {
+    if (plan) {
+      const planId = plan.identifier;
+      loadLocations(serviceClass, planId, fetchLocationsAction).catch(err => displayError(err));
+    }
+  }, []);
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx]: Is there ever a situation where plan is not in the store when this first mounts? Is there ever a scenario where the plan might change?

Seems like no, but just trying to think of edge cases :)

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx [hunk 16]
```diff
@@ -93,6 +94,14 @@ const UpdatePlan = (props: RouteComponentProps<RouteParams> & UpdatePlanProps) =
...
     ...propsForUpdatingPlans,
     ...(plan && { initialValues: getPlanFormValues(plan) }),
+    /** a renderProp prop. this tells the planForm; I will give you a component that knows of the plan you are displaying,
+     * the component will get jurisdictions associated with that plan and render them as links, what you(planForm)
+     * will do, is provide a child prop(a renderProp) that tells the mentioned component what other stuff you would wish
+     * displayed alongside the jurisdiction names links.
+     */
+    renderLocationNames: (
+      child?: (locationName: string, locationId: string, index: number) => JSX.Element
+    ) => <ConnectedPlanLocationNames child={child} plan={plan} />,
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/index.tsx]: Should we move this typing to it's own line so it can be exported and used everywhere it needs to be defined?

File: src/store/ducks/opensrp/locations/index.ts [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+/** creates actions that remove locations assigned to a plan from store
+ * @param {string} planId -  id of plans with the above assigned locations
+ */
+export const removeLocations = (planId: string): RemoveLocationsAction => {
+  return {
+    locationsByPlanId: { [planId]: [] },
+    planId,
+    type: REMOVE_LOCATIONS,
+  };
+};
```
Comment (cKellyDesign) [inline src/store/ducks/opensrp/locations/index.ts]: Nice 😎

File: src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx (hunk 14 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx]: done

File: src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx (hunk 15 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/PlanLocationNames/index.tsx]: no, this component is  only shown when you are editing plan, not creating it.

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx (hunk 16 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/index.tsx]: yeap.

File: src/components/forms/PlanForm/index.tsx (hunk 13 above)
Comment (p-netm) [reply src/components/forms/PlanForm/index.tsx]: Am afraid the other alternative would be to get rid of renderLocationNames props entirely and use the `PlanLocationNames` component from within `PlanForm`. having a nested renderProp is a consequence of passing the component responsible for rendering the locationNames as a prop.
@cKellyDesign @moshthepitt
